### PR TITLE
Added brave vpn connection manager interface and initial Windows impls

### PR DESCRIPTION
**Pull request by simonhong**
Added vpn connection manager interface
Added Initial impls for Windows OS.

fix https://github.com/brave/brave-browser/issues/15805

Security review: https://github.com/brave/security/issues/468

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

1. Build command line tool `vpntool`  in bc directory - `ninja -C ../out/Components vpntool`
2. Run ../out/Components/vpntool.exe with various options
3. Print connections `../out/Components/vpntool.exe --connections`
4. Print entries `../out/Components/vpntool.exe --entries`
5. Print devices `../out/Components/vpntool.exe --devices`
6. Create new entry `../out/Components/vpntool.exe --create --host_name=xxx --vpn_name=xxx --user_name=xxx --password=xxx`
7. Delete entry `../out/Components/vpntool.exe --remove --vpn_name=xxx`
8. Connect entry `../out/Components/vpntool.exe --connect --vpn_name=xxx`
9. Disconnect entry `../out/Components/vpntool.exe --disconnect --vpn_name=xxx`

### CODE REVIEW

File: components/brave_vpn/winvpntool.cc
Comment (bsclifton) [inline components/brave_vpn/winvpntool.cc:699]: s/passowrd/password 😄

File: components/brave_vpn/winvpntool.cc
Comment (bsclifton) [inline components/brave_vpn/winvpntool.cc:626]: Definitely cool to have this in here to show how the "magic number" is made 😄 Maybe we can put a comment over `Demo` that just explains:
```
// NOTE: This code is never called, but this is how the magic number set in
// brave_vpn::CreateEntry() is created for the `CustomIPSecPolicies` field
```

Comment (bsclifton) [reply components/brave_vpn/winvpntool.cc:626]: We could also add a command line arg for this which outputs the INI output actually 😄

**Comment by simonhong** [comment]
@bsclifton Ready to review! PTAL :)

**Comment by simonhong** [comment]
@deeppandya My PR will make some merge conflict to your PR (https://github.com/brave/brave-core/pull/7904).
Both PR added new `brave_vpn` target :)
I think all conflict will on happen on gn files.
The difference is using`browser` sub directory and `ENABLE_BRAVE_VPN`.
As brave vpn feature will only be used on browser side, I think we don't need to use `browser` sub directory.
and build flag name convension is `ENABLE_FOO`.

File: BUILD.gn
Comment (bsclifton) [inline BUILD.gn:500]: Mentioned in DM; but it doesn't execute this by default (at least with `Component`). I ran manually via `src/brave/vendor/depot_tools/ninja.exe -C src/out/Component vpntool`

How does `tools` group get built?

File: components/brave_vpn/brave_vpn_connection_info.h
Comment (bsclifton) [inline components/brave_vpn/brave_vpn_connection_info.h:17]: Maybe these field names could be more clear? ex:
```
std::string connection_name; 
std::string server_name; // or `hostname`
std::string username; 
std::string password;
```

File: components/brave_vpn/brave_vpn_connection_manager_win.cc
Comment (bsclifton) [inline components/brave_vpn/brave_vpn_connection_manager_win.cc:53]: This will be an interesting case; I'm guessing a user may get a different server name / username / password if they choose a different location. In this case, we would want to support overwriting the existing connection so it has the new information

File: components/brave_vpn/winvpntool.cc
Comment (bsclifton) [inline components/brave_vpn/winvpntool.cc:697]: The usage lines can change to `vpntool.exe` 😄 Love seeing the std::string / base methods used here for CLI processing

File: components/brave_vpn/winvpntool.cc
Comment (bsclifton) [inline components/brave_vpn/winvpntool.cc:679]: Returning immediately would make it so that you can't run multiple commands at once. I think this is OK; but you may want to get entries and connections in the same usage

**Review by bsclifton — COMMENTED** [review]
Comments left; great job here! 😄

File: components/brave_vpn/brave_vpn_connection_manager_win.cc
Comment (simonhong) [reply components/brave_vpn/brave_vpn_connection_manager_win.cc:53]: yeah, maybe this update could be done by delete and add ?

**Comment by simonhong** [comment]
@bsclifton Addressed. Thanks for review!

File: BUILD.gn
Comment (simonhong) [reply BUILD.gn:500]: and it would be good to build any target with `npm run build`.
Ex, `npm run build -- --build_target=vpntool`. I'll try this as a separate task.

File: components/brave_vpn/brave_vpn_connection_manager_win.cc
Comment (simonhong) [reply components/brave_vpn/brave_vpn_connection_manager_win.cc:53]: How about fill this api when we have concrete UX?

**Review by bsclifton — APPROVED** [review]

**Comment by bsclifton** [comment]
Approved, pending security review and also review by @deeppandya 😄

File: components/brave_vpn/brave_vpn_connection_manager_win.cc
Comment (bsclifton) [reply components/brave_vpn/brave_vpn_connection_manager_win.cc:53]: SGTM! 😄

File: components/brave_vpn/utils_win.cc
Comment (fmarier) [inline components/brave_vpn/utils_win.cc:220]: `RAS_MaxPhoneNumber + 1`

File: components/brave_vpn/utils_win.cc
Comment (fmarier) [inline components/brave_vpn/utils_win.cc:223]: `RAS_MaxDeviceType + 1`

File: components/brave_vpn/utils_win.cc
Comment (fmarier) [inline components/brave_vpn/utils_win.cc:224]: `RAS_MaxDeviceName + 1`

File: components/brave_vpn/utils_win.cc
Comment (fmarier) [inline components/brave_vpn/utils_win.cc:287]: Are we sure that's the right way to use this API? There is a deprecation notice at the top of https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-writeprivateprofilestringw:

> Note  This function is provided only for compatibility with 16-bit versions of Windows. Applications should store initialization information in the registry.

**Review by fmarier — CHANGES_REQUESTED** [review]
I left a few suggestions around buffer sizes. Memory management is easy to get wrong in C so I tend to be extra-careful around these, but let me know if there's a reason why these were done with bare constants.

File: components/brave_vpn/utils_win.cc
Comment (simonhong) [reply components/brave_vpn/utils_win.cc]: Changed to `RAS_MaxEntryName + 1`.

File: components/brave_vpn/utils_win.cc
Comment (simonhong) [reply components/brave_vpn/utils_win.cc]: Changed to `DNLEN + 1`

File: components/brave_vpn/utils_win.cc
Comment (simonhong) [reply components/brave_vpn/utils_win.cc]: Changed to `UNLEN + 1` and `PWLEN + 1`.

File: components/brave_vpn/utils_win.cc
Comment (simonhong) [reply components/brave_vpn/utils_win.cc]: Done.

File: components/brave_vpn/utils_win.cc
Comment (simonhong) [reply components/brave_vpn/utils_win.cc]: Done.

File: components/brave_vpn/utils_win.cc
Comment (simonhong) [reply components/brave_vpn/utils_win.cc]: Done

**Review by simonhong — COMMENTED** [review]
Addressed all comments for buf length.
and looking at https://github.com/brave/brave-core/pull/8920/files#r645919998

**Comment by simonhong** [comment]
Hmm, I got `Policy error again` with this PR. need to check

File: components/brave_vpn/utils_win.cc
Comment (simonhong) [reply components/brave_vpn/utils_win.cc:303]: @fmarier I assume above comments were written to recommend to use registry for storing application init data instead of `.ini` file because this `WritePrivateProfileString` api is used to edit ini file format.
We are using this api just to edit phonebook file(.pbk) because this phonebook file is `ini` format.
IMO, I think it's safe to use this api just for editing phonebook file(ini format). cc @bsclifton

File: components/brave_vpn/utils_win.cc
Comment (fmarier) [inline components/brave_vpn/utils_win.cc:62]: `constexpr`

File: components/brave_vpn/utils_win.cc
Comment (fmarier) [inline components/brave_vpn/utils_win.cc:82]: nit: We may as well make this one `const` so that it's not accidentally modified later.

File: components/brave_vpn/utils_win.cc
Comment (simonhong) [reply components/brave_vpn/utils_win.cc]: Done.

File: components/brave_vpn/utils_win.cc
Comment (simonhong) [reply components/brave_vpn/utils_win.cc]: Done.

File: components/brave_vpn/utils_win.cc
Comment (bsclifton) [reply components/brave_vpn/utils_win.cc:303]: Good question @fmarier - I think the risk is acceptable. This is definitely not a preferred API by any means... but if this gets removed we can always revisit (ex: provide a polyfill or replace with a `sed` equivalent). There is a [UWP implementation for VPN methods](https://docs.microsoft.com/en-us/uwp/api/windows.networking.vpn?view=winrt-20348) which I'm sure Microsoft would prefer be used but the integration is a lot heavier

Comment (fmarier) [reply components/brave_vpn/utils_win.cc:303]: > I think the risk is acceptable. This is definitely not a preferred API by any means... but if this gets removed we can always revisit (ex: provide a polyfill or replace with a sed equivalent).

If there's one company that knows how to maintain software/APIs for the long term, it's Microsoft, so if you're comfortable relying on what sounds like pre-Win95 API, then :+1:

**Review by fmarier — APPROVED** [review]

**Comment by bsclifton** [comment]
Thanks for the great review, @fmarier 😄👍

File: BUILD.gn
Comment (bridiver) [reply BUILD.gn:499]: This does not belong here. If you need to build it separately you can do that regardless of the location. It especially does not make sense as a general "tools" imo.

**Pull request merged**
